compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -724,6 +736,135 @@ fn find_opaque_ty_constraints_for_tait(tcx: TyCtxt<'_>, def_id: LocalDefId) -> T
     hidden.ty
 }
 
+#[instrument(skip(tcx), level = "trace", ret)]
+fn may_define_opaque_type<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    opaque_def_id: LocalDefId,
+    span: Span,
+) -> bool {
+    if tcx.is_descendant_of(opaque_def_id.to_def_id(), def_id.to_def_id()) {
+        // If the opaque type is defined in the body of a function, that function
+        // may constrain the opaque type since it can't mention it in bounds.
+        return true;
+    }
+
+    if tcx.is_closure(def_id.to_def_id()) {
```

Comment:
Should this only apply for closures, or also for inline consts? IOW, should we check `is_typeck_child` here?

Reply:
Is there a test for the ban on closure signatures? I saw tests/ui/type-alias-impl-trait/issue-63263-closure-return.rs is changed to accommodate it, but I missed the "fail" test.